### DISCUSSION TITLE
Knowing pins state :)

### DISCUSSION THREAD

**Discussion by jczic** [micropython/micropython#11182] [ESP32]
Hello to all!

I haven't found any way to know the current state of configuration of the pins.
Without disturbing a pin, the fact of declaring it without any arguments allows to access it easily: ```Pin(n)```
So we can know its high/low state.
But if we want to know its minimal configuration such as its IN/OUT or PULL state, how can we do it from MicroPython on ESP32 ?

I remember that Daniel Compora allowed this in the Pycom versions at the time.
( a special call on the pin returned a tuple of the configuration with IN/OUT, PULL UP/DOWN/None etc...)

I ask, but maybe this has already been proposed and refused because of compatibility concerns between ports or with Espressif's low-level APIs...

**Thanks to all the team anyway!** 👍🏻

**Comment by rkompass** [comment]
You have to use repr(pin1) if pin1 is the pin object and extract the info from the returned string.
What repr() returns at present seems to depend on the hardware, i.e. the MP port. 
Ideally it would be a string that contains all info to instantiate the pin object, perhaps the command itself.
So non-default options should be present, default options might be absent.
Recently an issue was raised addressing the need to harmonize the different representation formats: [https://github.com/micropython/micropython/issues/11055](https://github.com/micropython/micropython/issues/11055).

**Comment by jczic** [comment]
Hi @rkompass and thank you for your feedback.
I have several MicroPython firmware images for ESP32 but I always get **Pin(n)** from ```repr(Pin(n))``` or ```repr(Pin(n, Pin.IN, ...))```.
It's the same whether the pin is instantiated with specific parameters or simply obtained by default (knowing that pins are already initialized, such as for the UART repl).

And default uart for example, I have a format in output (without the pins though):
```python
>> u = UART(0)
>> repr(u)
"UART(0, baudrate=115201, bits=8, parity=None, stop=1, tx=-1, rx=-1, rts=-1, cts=-1, txbuf=256, rxbuf=256, timeout=0, timeout_char=0)"
```

This is the kind of return that would be nice with the pins :)

**Reply by kjm1102** [reply]
If I forget who is what I usually do
```
>>> led=Pin(18, Pin.OUT); led, led()
(Pin(18, mode=OUT), 0)
```

**Reply by jczic** [reply]
Yes, but it doesn't work on ESP32 port with last MicroPython firmwares :(

**Reply by kjm1102** [reply]
Huh, weird. 
I get (Pin(18, mode=OUT), 0) for v1.17.0 on a rpi pico but only (Pin(18), 0) for the same version on an esp32
